GOVERNANCE.md Proposal (#9)
zarr-developers/governance: GOVERNANCE.md

https://github.com/orgs/zarr-developers/teams/core-devs

Decisions should be made in accordance with the mission and values of the Zarr project.

Zarr uses a “consensus seeking” process for making decisions. The group
tries to find a resolution that has no open objections among core developers.
Core developers are expected to distinguish between fundamental objections to a
proposal and minor perceived flaws that they can live with, and not hold up the
decision-making process for the latter.  If no option can be found without
objections, the decision is escalated to the SC, which will itself use
consensus seeking to come to a resolution. In the unlikely event that there is
still a deadlock, the proposal will move forward if it has the support of a
simple majority of the SC.

Decisions (in addition to adding core developers and SC membership as above)
are made according to the following rules:

- **Minor documentation changes**, such as typo fixes, or addition / correction of a
  sentence, require approval by a core developer *and* no disagreement or requested
  changes by a core developer on the issue or pull request page (lazy

>   consensus). Core developers are expected to give "reasonable time" to others
>   to give their opinion on the pull request if they’re not confident others
>   would agree, where "reasonable time" is likely double-digit hours.

Reviewer: ```suggestion
  consensus). Core developers are expected to give "reasonable time" after approval and before merging for others
  to give their opinion on the pull request if they’re not confident others
  would agree, where "reasonable time" is likely one working day.
```

How about this?

Author: WFM

  to give their opinion on the pull request if they’re not confident others
  would agree.

- **Code changes and major documentation changes** require agreement by *one*
  core developer *and* no disagreement or requested changes by a core developer
  on the issue or pull-request page (lazy consensus). For all changes of this type,

>   core developers are expected to give “reasonable time” after approval and before

Reviewer 2: Would it be better to narrow down *reasonable* time here. I know this is very context-dependent, but maybe it would be good to clarify if we are talking about hours, days weeks or months.

Author: Would work for me, @meggart. Anything in particular you'd propose?

Reviewer 2: I am not sure since I am not contributing to the Python codebase myself. A suggestion would be that we require a minimum delay of 2 weeks since opening the PR and 3 days since approval before a pull request gets merged. However feel free to suggest different times, I just think it takes away uncertainty from the core-dev who is merging the PR.

Author: Pushed a first approximation for these times.

# Abstract

The purpose of this document is to formalize the governance process used by the
`zarr` project, to clarify how decisions are made and how the various
elements of our community interact.

This is a consensus-based community project. Anyone with an interest in the
project can join the community, contribute to the project design, and
participate in the decision making process. This document describes how that
participation takes place, how to find consensus, and how deadlocks are
resolved.

# Roles And Responsibilities

## The Community

> The zarr community consists of anyone using or working with the project

Reviewer 3: ```suggestion
The Zarr community consists of anyone using or working with the project
```

Author: Happy to accept this, but we probably should write down explicitly the expected capitalization somewhere. I feel like I've seen both.

in any way.

## Contributors

A community member can become a contributor by interacting directly with the
project in concrete ways, such as:

- proposing, discussing, or reviewing a change to the code, documentation, or specification
  via a GitHub pull request to any https://github.com/zarr-developers repository;
- reporting a GitHub issue

>   to any https://github.com/zarr-developers repository;

Reviewer 3: Any reason there's a linebreak here?

Author: Likely to minimize the diff from the previous document.

- discussing examples or usage issues under the #zarr tag on StackOverflow 
  (like "How do I do X in zarr?").

among other possibilities. Any community member can become a contributor, and
all are encouraged to do so. By contributing to the project, community members
can directly help to shape its future.

Contributors are encouraged to read the [contributing guide](http://zarr.readthedocs.io/en/stable/contributing.html)
for information about contributing to Zarr.

## Core developers

Core developers are community members that have demonstrated continued
commitment to the project through ongoing contributions. They
have shown they can be trusted to maintain zarr with care. Becoming a
core developer allows contributors to merge approved pull requests, cast votes
for and against merging a pull-request, and be involved in deciding major
changes to the API, and thereby more easily carry on with their project related
activities. Core developers appear as organization members on the zarr
[GitHub organization](https://github.com/orgs/zarr-developers/people) and are on our

> [@zarr-developers/core-devs](https://github.com/orgs/zarr-developers/teams/core-devs) GitHub team. Core

Reviewer 4: I don't think this link is publicly accessible. It appears you need to (1) have a GitHub account and (2) either be a member of the `zarr-developers` org or the `core-devs` team to view the page.

We could consider removing the `core-devs` team link and just stick with the `zarr-developers` org members. Or if we're feeling more ambitious we could add additional content to https://zarr-developers.github.io which lists core developers. Perhaps using something along the lines of https://jupyter.org/about as a model

Author: Having a page sounds good. Someone want to give it a shot?

developers are expected to review code contributions while adhering to the
[core developer guide](CORE_DEV_GUIDE.md). New core developers can be nominated
by any existing core developer, and for details on that process see our core

> developer guide.

Reviewer 5: If we included the details of how one can become a contributor and steering council member, then this document would be more self contained. I would be in favor of adding these details.

Author: I looked at a few core-dev documents and the modifications were going to be pretty substantial. I like the idea of working all of the content into the same file and perhaps starting with less and building up over time. Happy to get a start on that if there are no objections.

Author: Strike that. Was easier to get `CORE_DEV_GUIDE.md` as it stood into a form that was reviewable. Opinions still welcome on merging it into `GOVERNANCE.md`.

## Steering Council

The Steering Council (SC) members are core developers who have additional
responsibilities to ensure the smooth running of the project. SC members are
expected to participate in strategic planning, approve changes to the
governance model, and make decisions about funding granted to the project
itself. (Funding to community members is theirs to pursue and manage). The
purpose of the SC is to ensure smooth progress from the big-picture
perspective. Changes that impact the full project require analysis informed by
long experience with both the project and the larger ecosystem. When the core
developer community (including the SC members) fails to reach such a consensus
in a reasonable timeframe, the SC is the entity that resolves the issue.

Members of the steering council also have the "owner" role within the [zarr-developers GitHub organization](https://github.com/zarr-developers/)

> and are ultimately responsible for managing the zarr-developers GitHub account, the [@zarr_dev](https://twitter.com/zarr_dev)

Reviewer 3: ```suggestion
and are ultimately responsible for managing the [zarr-developers](https://github.com/zarr-developers) GitHub account, the [@zarr_dev](https://twitter.com/zarr_dev)
```

for completeness :)

twitter account, the [zarr website]( https://zarr-developers.github.io), and other similar zarr owned resources.

The steering council is currently fixed in size to five members. This number may be increased in
the future, but will always be an odd number to ensure a simple majority vote outcome
is always possible. The initial steering council of zarr consists of

> * [Ryan Abernathy](https://github.com/rabernat)

Reviewer 5: ```suggestion
* [Ryan Abernathey](https://github.com/rabernat)
```

of the zarr project.

zarr uses a “consensus seeking” process for making decisions. The group
tries to find a resolution that has no open objections among core developers.
Core developers are expected to distinguish between fundamental objections to a
proposal and minor perceived flaws that they can live with, and not hold up the
decision-making process for the latter.  If no option can be found without
objections, the decision is escalated to the SC, which will itself use
consensus seeking to come to a resolution. In the unlikely event that there is
still a deadlock, the proposal will move forward if it has the support of a
simple majority of the SC.

Decisions (in addition to adding core developers and SC membership as above)
are made according to the following rules:

- **Minor documentation changes**, such as typo fixes, or addition / correction of a
  sentence, require approval by a core developer *and* no disagreement or requested
  changes by a core developer on the issue or pull request page (lazy
  consensus). Core developers are expected to give "reasonable time" to others
  to give their opinion on the pull request if they’re not confident others

>   would agree, where "reasonable time" is likely double-digit hours.

Reviewer 3: I'd advocate for hard time and explicit time limits everywhere, just to remove the cognitive load on people making contributions.

Author: WFM. Suggestion?

  sentence, require approval by a core developer *and* no disagreement or requested
  changes by a core developer on the issue or pull request page (lazy
  consensus). Core developers are expected to give "reasonable time" after approval and before merging for others
  to give their opinion on the pull request if they’re not confident others
  would agree, where "reasonable time" is likely one working day.

- **Code changes and major documentation changes** require agreement by *one*
  core developer *and* no disagreement or requested changes by a core developer
  on the issue or pull-request page (lazy consensus). For all changes of this type,
  core developers are expected to give "reasonable time" after approval and before
  merging for others to weigh in on the pull request in its final state, where
  "reasonable time" is likely a few working days.

- **Changes to the Python API principles** require a dedicated issue on our
  [issue tracker](https://github.com/zarr-developers/zarr-python/issues) and follow the
  decision-making process outlined above, though "reasonable time" is likely extended

>   to at least a week.

Reviewer 6: Should we have a similar section for spec and other repos? Or would it make sense to generalize this to cover all (even non-Python) repos?

Author: Something like 15fc40f11a955c67ce001ad671845cfdfc70ad80 ?

Reviewer: FWIW I think we'll need a separate discussion to figure out the process for the specs repo, it's a bit different from the code repos. E.g., any changes to specs can impact on multiple implementations, and so will need higher level of consensus. 

How about we:

* Describe a general process that applies to all code repositories in the zarr-developers organisation.
* Say that a separate process applies to the zarr-specs repository, to be defined.

Author: @alimanfoo _et al_. : please see d8f82d3377bc8c228b8890e3073744ac90e853cd
